code/_onclick/hud/radial.dm

```diff
@@ -0,0 +1,311 @@
+	if(tooltips)
+		closeToolTip(usr)
+
+/obj/screen/radial/slice/Click(location, control, params)
+	if(usr.client == parent.current_user)
+		if(next_page)
+			parent.next_page()
+		else
+			parent.element_chosen(choice,usr)
+
+/obj/screen/radial/center
+	name = "Close Menu"
+	icon_state = "radial_center"
+
+/obj/screen/radial/center/MouseEntered(location, control, params)
+	. = ..()
+	icon_state = "radial_center_focus"
+
+/obj/screen/radial/center/MouseExited(location, control, params)
+	. = ..()
+	icon_state = "radial_center"
+
+/obj/screen/radial/center/Click(location, control, params)
+	if(usr.client == parent.current_user)
+		parent.finished = TRUE
+
+/datum/radial_menu
+	var/list/choices = list() //List of choice id's
+	var/list/choices_icons = list() //choice_id -> icon
+	var/list/choices_values = list() //choice_id -> choice
+	var/list/page_data = list() //list of choices per page
+
+
+	var/selected_choice
+	var/list/obj/screen/elements = list()
+	var/obj/screen/radial/center/close_button
+	var/client/current_user
+	var/atom/anchor
+	var/image/menu_holder
+	var/finished = FALSE
+	var/datum/callback/custom_check_callback
+	var/next_check = 0
+	var/check_delay = DEFAULT_CHECK_DELAY
+
+	var/radius = 32
+	var/starting_angle = 0
+	var/ending_angle = 360
+	var/zone = 360
+	var/min_angle = 45 //Defaults are setup for this value, if you want to make the menu more dense these will need changes.
+	var/max_elements
+	var/pages = 1
+	var/current_page = 1
+
+	var/hudfix_method = TRUE //TRUE to change anchor to user, FALSE to shift by py_shift
+	var/py_shift = 0
+	var/entry_animation = TRUE
+
+//If we swap to vis_contens inventory these will need a redo
+/datum/radial_menu/proc/check_screen_border(mob/user)
+	var/atom/movable/AM = anchor
+	if(!istype(AM) || !AM.screen_loc)
+		return
+	if(AM in user.client.screen)
+		if(hudfix_method)
+			anchor = user
+		else
+			py_shift = 32
+			restrict_to_dir(NORTH) //I was going to parse screen loc here but that's more effort than it's worth.
+
+//Sets defaults
+//These assume 45 deg min_angle
+/datum/radial_menu/proc/restrict_to_dir(dir)
+	switch(dir)
+		if(NORTH)
+			starting_angle = 270
+			ending_angle = 135
+		if(SOUTH)
+			starting_angle = 90
+			ending_angle = 315
+		if(EAST)
+			starting_angle = 0
+			ending_angle = 225
+		if(WEST)
+			starting_angle = 180
+			ending_angle = 45
+
+/datum/radial_menu/proc/setup_menu(use_tooltips)
+	if(ending_angle > starting_angle)
+		zone = ending_angle - starting_angle
+	else
+		zone = 360 - starting_angle + ending_angle
+
+	max_elements = round(zone / min_angle)
+	var/paged = max_elements < choices.len
+	if(elements.len < max_elements)
+		var/elements_to_add = max_elements - elements.len
+		for(var/i in 1 to elements_to_add) //Create all elements
+			var/obj/screen/radial/slice/new_element = new /obj/screen/radial/slice
+			new_element.tooltips = use_tooltips
+			new_element.parent = src
+			elements += new_element
+
+	var/page = 1
+	page_data = list(null)
```

Comment:
Why isn't this just a list()?

Reply:
Pretty sure that's done to avoid bad index runtimes due to how it's done. See lines 134-137 and 147.